C/eval.h

```diff
@@ -25,23 +30,25 @@
  *               input == NULL or UWORD input[ROUND_UWORD(inputSize)];
  *               if 'dag[len]' represents a Simplicity expression with primitives then 'NULL != env';
  */
-bool evalTCOExpression( bool *evalSuccess, UWORD* output, size_t outputSize, const UWORD* input, size_t inputSize
+bool evalTCOExpression( bool *evalSuccess, bool *evalCheck, UWORD* output, size_t outputSize, const UWORD* input, size_t inputSize
                       , const dag_node* dag, type* type_dag, size_t len, const txEnv* env
                       );
 
 /* Run the Bit Machine on the well-typed Simplicity program 'dag[len]'.
  *
- * If static analysis results determines the bound on memory allocation requirements exceed the allowed limits,
- * set '*evalSuccess' to 'false' and return 'true'.
  * If malloc fails, return 'false', otherwise return 'true'.
- * If during execution an 'assertr' or 'assertl' combinator fails, '*evalSuccess' is set to 'false'
- * Otherwise '*evalSuccess' is set to 'true'.
+ * If static analysis results determines the bound on memory allocation requirements exceed the allowed limits,
+ * set '*evalPass' to 'false'.
+ * If during execution an 'assertr' or 'assertl' combinator fails, '*evalPass' is set to 'false'.
+ * If after execution the anti-DOS checks fail, '*evalPass' is set to 'false'.
+ * Otherwise '*evalPass' is set to 'true'.
  *
- * Precondition: NULL != evalSuccess
+ * Precondition: NULL != evalPass
  *               dag_node dag[len] and 'dag' is well-typed with 'type_dag' of type 1 |- 1;
  *               if 'dag[len]' represents a Simplicity expression with primitives then 'NULL != env';
  */
-static inline bool evalTCOProgram(bool *evalSuccess, const dag_node* dag, type* type_dag, size_t len, const txEnv* env) {
-  return evalTCOExpression(evalSuccess, NULL, 0, NULL, 0, dag, type_dag, len, env);
+static inline bool evalTCOProgram(bool *evalPass, const dag_node* dag, type* type_dag, size_t len, const txEnv* env) {
+  bool evalSuccess;
+  return evalTCOExpression(&evalSuccess, evalPass, NULL, 0, NULL, 0, dag, type_dag, len, env);
```

Comment:
In 2d58ec81f4a848b7f7083e3e85ba85e78545b7b6:

I think it'd be clearer for `evalTCOExpression` to not take the `evalSuccess` argument at all, since it seems like you always ignore it.

Reply:
There is an API design here that I'm struggling with.  In some sense we should always fail when the anti-DOS tests fail.  On the other hand, we currently have tests (well currently a test) for sha256-compression that is unpruned, and if we wanted to test it on random input, it must be unpruned.

So my compromise is that the *program* evaluator (recall a program is an expression from unit to unit) always does the anti-DOS checks, while the lower level expression evaluator lets you optionally run or ignore the anti-DOS checks.

`evalTCOExpression`'s `evalSuccess` parameter is used in `test.c`. 

There is probably a million ways we could handle this success, check, pass, boolean logic, and I'm open to alternatives.  Currently `evalCheck` can only be true if `evalSuccess` is true, which is why `evalSuccess` is ignored in the above lines.

Reply:
Done.